### PR TITLE
Add x86 support for optee driver

### PR DESCRIPTION
**Pull request by jingdlu**
This driver only supports optee-os x86 3.13 version.

### CODE REVIEW

**Comment by jingdlu** [comment]
@JeevakaPrabu Could you please review this PR?

File: drivers/tee/optee/Kconfig [hunk 1]
```diff
@@ -15,3 +15,34 @@ config OPTEE_SHM_NUM_PRIV_PAGES
 	help
 	  This sets the number of private shared memory pages to be
 	  used by OP-TEE TEE driver.
+
+config OPTEE_BENCHMARK
```
Comment (James-Zzz) [inline drivers/tee/optee/Kconfig:19]: Do we need to add comment to indicate that benchmark is not supported in x86?

File: drivers/tee/optee/core.c [hunk 2]
```diff
@@ -27,6 +36,102 @@
...
+
+#if defined(CONFIG_OPTEE_IVSHMEM)
+#define PCI_DRV_NAME	"optee-ivshmem"
+
+#define IVPOSITION_OFF	0x08
+#define DOORBELL_OFF	0x0C
+
+#define OPTEE_SHM_QUEUE_SIZE	64
+
+#define OPTEE_HANDLE_DONE		0xa5a5a5a5
```
Comment (James-Zzz) [inline drivers/tee/optee/core.c:64]: Suggest to keep same indent as line 62 (some tab) for line 64 and 66 (two tabs)

File: drivers/tee/optee/core.c [hunk 3]
```diff
@@ -27,6 +36,102 @@
...
+} __packed;
+
+struct optee_smc_ring {
+	u16 head;
+	u16 tail;
+	u16 ring[OPTEE_SHM_QUEUE_SIZE];
+} __packed;
+
+static struct optee_smc_args *g_smc_args = NULL;
+static struct optee_smc_ring *smc_avail_ring = NULL;
```
Comment (James-Zzz) [inline drivers/tee/optee/core.c:122]: In line 90 and 121, global variable for file scope contains prefix "g_", suggest to keep same style for line 122/123/124

File: drivers/tee/optee/core.c [hunk 4]
```diff
@@ -763,13 +1109,331 @@ static int optee_probe(struct platform_device *pdev)
...
+	if (ret) {
+		pr_err("ivshmem unable to enable msix: %d\n", ret);
+		goto free_names;
+	}
+
+	for (i = 0; i < ivshmem_dev->nvectors; i++) {
+		snprintf(ivshmem_dev->msix_names[i], sizeof(*ivshmem_dev->msix_names),
+				"%s-%d", PCI_DRV_NAME, i);
+
+		ret = request_irq(ivshmem_dev->msix_entries[i].vector,
```
Comment (James-Zzz) [inline drivers/tee/optee/core.c:1205]: Request_irq use vector number here, which means we try to map IRQ number equals to Vector number, right?
request_irq inner function will allocate vector itself, so the will the allocated vector matches the vector number we provide?

**Review by James-Zzz — APPROVED** [review]

File: drivers/tee/optee/core.c (hunk 4 above)
Comment (jingdlu) [reply drivers/tee/optee/core.c:1205]: Here, ivshmem_dev->msix_entries[i].vector should be IRQ number.

File: drivers/tee/optee/Kconfig (hunk 1 above)
Comment (jingdlu) [reply drivers/tee/optee/Kconfig:19]: Will comment it for x86.

File: drivers/tee/optee/core.c (hunk 2 above)
Comment (jingdlu) [reply drivers/tee/optee/core.c:64]: Will change it.

**Review by JeevakaPrabu — APPROVED** [review]

**Pull request merged**